### ISSUE TITLE
How can I implement a custom operations with state resource?

### ISSUE THREAD

**Issue by MoFHeka** [google/jax#22412]
For example, there is a lookup table backend, and this operation need to get the table handle to reach it. This feature is implemented in tensorflow by tf.resource.

Please:

- [x] Check for duplicate requests.
- [x] Describe your goal, and if possible provide a code snippet with a motivating example.

**Comment by justinjfu** [comment]
Yes, `jax.experimental.io_callback` would be the recommended way to implement this. You may also want to look at `jax.dlpack` to avoid copying arrays and using the underlying buffer directly.

**Comment by MoFHeka** [comment]
@justinjfu What is the right way to achieve optimal performance using jax.experimental.io_callback and jax.dlpack? Are there any examples you can refer to?

**Comment by superbobry** [comment]
All callbacks copy inputs to the host and copy outputs to the device, so there is at least two copies if you are running on GPU/TPU.

On CPU there is a single copy of callback outputs into the pre-allocated XLA output buffers. I have a prototype which allows writing zero-copy CPU callbacks for #20701, but it is not quite ready yet.

That said, it sounds like you are asking about something else. Would the table handle be shared by all invocations of the callback? How do you currently pass it in?

**Comment by MoFHeka** [comment]
@superbobry All callbacks copy twice sounds unacceptable.
We're migrating code from TensorFlow into Jax. And we found custom operator in Jax was unable to passed a resource handle input like Tensorflow ResourceBase. 
Now we have several custom lookup table implementation, we usually initialize them as a tf.resource object just like tf.LookupTable. And then we pass these to tf.resource custom operator, such as lookup, insert, export. 
But I found this path is impossible when using Jax?

**Comment by superbobry** [comment]
Can you clarify if you want all invocations of the callback to use the same resource?

**Comment by MoFHeka** [comment]
Yes, of course. This resource is a hash table, and we need to use it when forwards and backwards in every training steps. We lookup huge size embedding from this hash table when forwards, and insert the updated embedding when backwards.在 2024年7月16日，17:27，Sergei Lebedev ***@***.***> 写道：﻿
Can you clarify if you want all invocations of the callback to use the same resource?

**Comment by MoFHeka** [comment]
@superbobry Any progress?

**Comment by superbobry** [comment]
I'm afraid the problem you are trying to solve is still not quite clear to me. I understand the use-case, but I cannot advise unless you share the code (even if a sketch) of the JAX part of the custom operation.

From your description so far it sound like you can probably pass the pointer to the hash table via `backend_config=`?

**Comment by MoFHeka** [comment]
I found this document and I would try it later.
https://jax--22095.org.readthedocs.build/en/22095/ffi/ffi.html
